app/handler/provider_complaint.py

```diff
@@ -0,0 +1,233 @@
+            if current_part == 3:
+                return part
+        return None
+
+    @classmethod
+    def name(cls):
+        return "hotmail"
+
+
+def handle_hotmail_complaint(message: Message) -> bool:
+    return handle_complaint(message, TransactionalHotmailOrigin())
+
+
+def handle_yahoo_complaint(message: Message) -> bool:
+    return handle_complaint(message, TransactionalYahooOrigin())
+
+
+def find_alias_with_address(address: str) -> Optional[Alias]:
+    return (
+        Alias.get_by(email=address)
+        or DeletedAlias.get_by(email=address)
+        or DomainDeletedAlias.get_by(email=address)
+    )
+
+
+def handle_complaint(message: Message, origin: ProviderComplaintOrigin) -> bool:
+    original_message = origin.get_original_message(message)
+
+    try:
+        _, to_address = parse_full_address(
+            get_header_unicode(original_message[headers.TO])
+        )
+        _, from_address = parse_full_address(
+            get_header_unicode(original_message[headers.FROM])
+        )
+    except ValueError:
+        saved_file = save_email_for_debugging(message, "FromParseFailed")
+        LOG.w(f"Cannot parse from header. Saved to {saved_file or 'nowhere'}")
+        return False
+
+    user = User.get_by(email=to_address)
+    if user:
+        LOG.d(f"Handle provider {origin.name()} complaint for {user}")
+        report_complaint_to_user(user, origin)
+        return True
+
+    alias = find_alias_with_address(from_address)
+    # the email is during a reply phase, from=alias and to=destination
+    if alias:
+        LOG.i(
+            f"Complaint from {origin.name} during reply phase {alias} -> {to_address}, {user}"
+        )
+        report_complaint_to_user_in_reply_phase(alias, to_address, origin)
+        store_provider_complaint(alias, message)
+        return True
+
+    contact = Contact.get_by(reply_email=from_address)
+    if contact:
+        alias = contact.alias
+    else:
+        alias = find_alias_with_address(to_address)
+
+    if not alias:
+        LOG.e(
+            f"Cannot find alias from address {to_address} or contact with reply {from_address}"
+        )
+        return False
+
+    report_complaint_to_user_in_forward_phase(alias, origin)
+    return True
+
+
+def report_complaint_to_user_in_reply_phase(
+    alias: Alias, to_address: str, origin: ProviderComplaintOrigin
+):
+    capitalized_name = origin.name().capitalize()
+    send_email_with_rate_control(
+        alias.user,
+        f"{ALERT_COMPLAINT_REPLY_PHASE}_{origin.name()}",
+        alias.user.email,
+        f"Abuse report from {capitalized_name}",
+        render(
+            "transactional/provider-complaint-reply-phase.txt.jinja2",
+            user=alias.user,
+            alias=alias,
+            destination=to_address,
+            provider=capitalized_name,
+        ),
+        max_nb_alert=1,
+        nb_day=7,
+    )
+
+
+def report_complaint_to_user(user: User, origin: ProviderComplaintOrigin):
```

Comment:
Nits: can we rename this method to `report_complaint_to_user_in_transactional_phase`?

Reply:
Changed